plugins/com.google.cloud.tools.eclipse.appengine.newproject.maven/src/com/google/cloud/tools/eclipse/appengine/newproject/maven/MavenAppEngineStandardWizardPage.java

```diff
@@ -323,4 +329,69 @@ public void modifyText(ModifyEvent event) {
       checkFlipToNext();
     }
   }
+
+  /**
+   * Auto-fills javaPackageField as "groupId.artifactId" (or "groupId" if artifactId is empty),
+   * only when 1) javaPackageField is empty; or 2) the field matches previous auto-fill before
+   * ID modification.
+   */
+  private final class AutoPackageNameSetterOnGroupIdChange implements VerifyListener {
+    @Override
+    public void verifyText(VerifyEvent event) {
+      // Below explains how to get text after modification:
+      // http://stackoverflow.com/questions/32872249/get-text-of-swt-text-component-before-modification
+      String newGroupId =
+          getGroupId().substring(0, event.start) + event.text + getGroupId().substring(event.end);
+
+      String oldPackageName = suggestPackageName(getGroupId(), getArtifactId());
+      String newPackageName = suggestPackageName(newGroupId, getArtifactId());
+      adjustPackageName(oldPackageName, newPackageName);
+    }
+  }
+
+  /**
+   * See {@link AutoPackageNameSetterOnGroupIdChange}.
+   */
+  private final class AutoPackageNameSetterOnArtifactIdChange implements VerifyListener {
+    @Override
+    public void verifyText(VerifyEvent event) {
+      String newArtifactId = getArtifactId().substring(0, event.start)
+          + event.text + getArtifactId().substring(event.end);
+
+      String oldPackageName = suggestPackageName(getGroupId(), getArtifactId());
+      String newPackageName = suggestPackageName(getGroupId(), newArtifactId);
+      adjustPackageName(oldPackageName, newPackageName);
+    }
+  }
+
+  /**
+   * See {@link AutoPackageNameSetterOnGroupIdChange#verifyText(VerifyEvent)}.
+   */
+  private void adjustPackageName(String oldPackageName, String newPackageName) {
+    if (getPackageName().isEmpty() || getPackageName().equals(oldPackageName)) {
+      javaPackageField.setText(newPackageName);
+    }
+  }
+
+  /**
+   * Helper function returning a suggested package name based on groupId and artifactId. It
+   * does basic string filtering/manipulation for valid Java package names, which is not perfect.
+   * However, users will be alerted of any slipping errors in naming by {@link #validatePage}.
+   */
+  @VisibleForTesting
+  protected static String suggestPackageName(String groupId, String artifactId) {
```

Comment:
white listing would be safer than blacklisting, but I already wrote this in the new project wizard, let me look. Yep, I did. see com.google.cloud.tools.eclipse.appengine.newproject.JavaPackageValidator.